crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,590 @@
+//! This module implements user space page fault handling with the `userfaultfd` ("uffd") system call on Linux.
+//!
+//! Handling page faults for memory accesses in regions relating to WebAssembly instances
+//! enables the runtime to protect guard pages in user space rather than kernel space (i.e. without `mprotect`).
+//!
+//! Additionally, linear memories can be lazy-initialized upon first access.
+//!
+//! Handling faults in user space is slower than handling faults in the kernel. However,
+//! in use cases where there is a high number of concurrently executing instances, handling the faults
+//! in user space requires rarely changing memory protection levels.  This can improve concurrency
+//! by not taking kernel memory manager locks and may decrease TLB shootdowns as fewer page table entries need
+//! to continually change.
+//!
+//! Here's how the `uffd` feature works:
+//!
+//! 1. A user fault file descriptor is created to monitor specific areas of the address space.
+//! 2. A thread is spawned to continually read events from the user fault file descriptor.
+//! 3. When a page fault event is received, the handler thread calculates where the fault occurred:
+//!    a) If the fault occurs on a linear memory page, it is handled by either copying the page from
+//!       initialization data or zeroing it.
+//!    b) If the fault occurs on a guard page, the protection level of the guard page is changed to
+//!       force the kernel to signal SIGBUS on the next retry. The faulting page is recorded so the
+//!       protection level can be reset in the future.
+//! 4. Faults to address space relating to an instance may occur from both Wasmtime (e.g. instance
+//!    initialization) or from WebAssembly code (e.g. reading from or writing to linear memory),
+//!    therefore the user fault handling must do as little work as possible to handle the fault.
+//! 5. When the pooling allocator is dropped, it will drop the memory mappings relating to the pool; this
+//!    generates unmap events for the fault handling thread, which responds by decrementing the mapping
+//!    count. When the count reaches zero, the user fault handling thread will gracefully terminate.
+//!
+//! This feature requires a Linux kernel 4.11 or newer to use.
+
+use super::{InstancePool, MemoryPool};
+use crate::instance::Instance;
+use anyhow::{bail, Context, Result};
+use std::thread;
+use userfaultfd::{Event, FeatureFlags, IoctlFlags, Uffd, UffdBuilder};
+use wasmtime_environ::{entity::EntityRef, wasm::DefinedMemoryIndex, MemoryInitialization};
+
+const WASM_PAGE_SIZE: usize = wasmtime_environ::WASM_PAGE_SIZE as usize;
+
+fn decommit(addr: *mut u8, len: usize) -> Result<()> {
+    if len == 0 {
+        return Ok(());
+    }
+
+    unsafe {
+        // On Linux, this is enough to cause the kernel to initialize the pages to 0 on next access
```

Comment:
I think this comment may want to be updated from the other comment that looks like this as well. By my reading it's not enough to just zero the pages, but we also need to guarantee that all future faults to these pages to through to uffd, right? So this is technically both zero-ing and ensuring that faults get routed right?